VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,658 @@
+
+- The NVD https://nvd.nist.gov/vuln/data-feeds#cpeMatch defines CPE ranges as
+  lists of version start and end either including or excluding the start or end
+  version. And also provides a concrete enumeration of the available ranges as
+  a daily feed.
+
+- The version 5 of the NVD CVE JSON data format at
+  https://github.com/CVEProject/cve-schema/blob/master/schema/v5.0/CVE_JSON_5.0.schema#L303
+  defines version ranges with a starting version, a versionType, and an upper
+  limit for the version range as lessThan or lessThanOrEqual. Or an enumeration
+  of versions. The versionType is defined as ``"The version numbering system
+  used for specifying the range. This defines the exact semantics of the
+  comparison (less-than) operation on versions, which is required to understand
+  the range itself"``. A versionType ressembles closely the Package URL package "type".
+
+- The OSSF OSV schema https://ossf.github.io/osv-schema/ defines vulnerable
+  ranges with version events using "introduced" and "limit" fields and an
+  enumeration of all the versions in these ranges, except for semver-based
+  versions. A range may be ecosystem-specific based on a provided package
+  "ecosystem" value that ressembles closely the Package URL package "type".
+
+
+The way two versions are compared as equal, lesser or greater is a closely
+related topic:
+
+- Each package ecosystem may have evolved its own peculiar version string
+  conventions, semantics and comparison procedure.
+
+- For instance, ``semver`` is a prominent specification in this domain but this is
+  just one of the many ways to structure a version string.
+
+- Debian, RPM, PyPI, Rubygems, and Composer have their own subtly different
+  approach on how to determine which version is greater or lesser.
+
+
+Solution
+---------
+
+A solution to the many version range syntaxes is to design a new notation to
+unify them all with:
+
+- a mostly universal and minimalist, compact notation to express version ranges
+  from many different package types and ecosystems.
+
+- the package type-specific definitions to normalize existing range expressions
+  to this common notation.
+
+- the designation of which algorithm or procedure to use when comparing two
+  versions such that it is possible to resolve if a version is within or
+  outside of a version range.
+
+We call this solution "version range specifier" or "vers" and it is described
+in this document.
+
+
+Version range specifier
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A version range specifier (aka. "vers") is a URI string using the ``vers``
+URI-scheme with this syntax::
+
+   vers:<versioning-scheme>/<version-constraint>|<version-constraint,version-constraint>,...
+
+For example to define a set of versions that contains either version ``1.2.3``,
+or any versions greater than or equal to ``2.0.0`` but less than ``5.0.0`` using
+the ``node-semver`` versioning scheme, the version range specifier will be::
+
+    vers:npm/1.2.3|>=2.0.0,<5.0.0
```

Comment:
I may have missed this but how should an implementation react to an impossible range? Something such as `vers:npm/<1.0.0,>5.0.0`? Does the result not match or should an impossible range be invalid?

Reply:
This is an excellent point... The proposal of @oliverchang that I am push in a few echews this entirely and makes it impossible to have impossible or overlapping range constraints and is beautifully simple and obvious.
That said, this make me thing that I shall add that constraints should be unique.

Reply:
I added the dedup and mentioned that ranges would always be non-overlapping